lib/src/source/git.dart

```diff
@@ -253,9 +302,22 @@ class GitSource extends CachedSource {
   /// by [id] on the effective ref of [id].
   ///
   /// This assumes that the canonical clone already exists.
-  Future<String> _getRev(PackageId id) {
-    return git.run(["rev-list", "--max-count=1", _getEffectiveRef(id)],
-        workingDir: _repoCachePath(id)).then((result) => result.first);
+  Future<String> _getRev(PackageId id) async {
+    var ref = _getEffectiveRef(id);
```

Comment:
Yeah, I forgot that `PackageId` doesn't include the original version constraint because it's not part of the description. This is going to be trickier than I thought.

The underlying issue is that Git IDs are no longer context-independent in the way they used to be; we don't know whether one was created due to new-style version resolution or old-style looking at `HEAD`. Git does have a means of "resolving" package IDs, but because `getVersions()` returns `Pubspec`s rather than `PackageId`s, those resolved IDs don't make their way into the solver.

This probably calls for a full refactor of `getVersions()` that changes the signature to `List<PackageId> getVersions(PackageRef ref)` and moves some caching logic into the hosted source. This is out of the scope of what you're doing, though, so I'll implement it myself; hold tight until then.